frontend/src/scenes/pipeline/AppCode.tsx

```diff
@@ -0,0 +1,85 @@
+import { LemonButton, LemonTabs, Spinner } from '@posthog/lemon-ui'
+import { useActions, useValues } from 'kea'
+import { CodeEditor } from 'lib/components/CodeEditors'
+
+import { PluginInstallationType } from '~/types'
+
+import { appsCodeLogic } from './appCodeLogic'
+import { appsManagementLogic } from './appsManagementLogic'
+
+export function AppCode({
+    pluginId,
+    pluginType,
+}: {
+    pluginId: number
+    pluginType: PluginInstallationType
+}): JSX.Element {
+    const logic = appsCodeLogic({ pluginId })
+    const { currentFile, fileNames, pluginSource, pluginSourceLoading, editMode } = useValues(logic)
+    const { setCurrentFile, updateCode, setEditMode, saveCode, fetchPluginSource } = useActions(logic)
+    const { canGloballyManagePlugins } = useValues(appsManagementLogic)
+
+    if (pluginSourceLoading) {
+        return <Spinner />
+    }
+
+    const canEdit = canGloballyManagePlugins && pluginType === PluginInstallationType.Source
+    const formattedFileContent = pluginSource[currentFile]
+
+    const content = (
+        <>
+            {canEdit && (
+                <div className="flex items-center gap-2">
+                    {editMode ? (
+                        <>
+                            <LemonButton
+                                type="secondary"
+                                onClick={() => {
+                                    fetchPluginSource()
+                                }}
+                            >
+                                Reset
+                            </LemonButton>
+                            <LemonButton
+                                type="primary"
+                                onClick={() => {
+                                    saveCode()
+                                }}
+                            >
+                                Save
+                            </LemonButton>
+                        </>
+                    ) : (
+                        <LemonButton type="secondary" onClick={() => setEditMode(true)}>
+                            Edit
+                        </LemonButton>
+                    )}
+                </div>
+            )}
+            <CodeEditor
+                // TODO: For some reason the editor doesn't keep the cursor focus and then I can type only one char and need to click again
```

Comment:
is there an easy way I can fix this. I saw that in the old UI it's a form and then the onChange is there

Reply:
There are two possibilities:
1. Something else steals the focus (unlikely)
2. This `CodeEditor` component is rendered with a different identity when its `value` changes. In my experience, this happens, but I don't immediately see why in this case. Might be something with `LemonTabs`? The React Dev Tools' have a tree view which is really handy for tracking down re-renders like this

Reply:
somewhat related: I'm thinking we only want a single app code edited at once probably - so a pop-up modal for edit and in here it's readOnly?

Reply:
> The React Dev Tools' have a tree view which is really handy for tracking down re-renders like this

you can show me on Monday :)